oeps/processes/oep-0055-proc-project-maintainers.rst

```diff
@@ -0,0 +1,450 @@
+"well-maintained," and feedback looks to ensure accountability and
+continuous improvement of our processes.
+
+Specification
+=============
+
+Definitions
+-----------
+
+The term **repository** will mean a repository on the GitHub social
+coding platform.
+
+The term **component** will typically mean the unit of software housed
+in a single GitHub repository.  Exceptions will be made in the case of
+edx-platform where there may be multiple distinguishable components in
+that single repository.  Components will typically, but not always, be
+hosted at the `openedx GitHub organization`_. Components will be
+either "part of" or "critical to" the open-source platform.
+
+The term **maintainer** will mean a named person or persons
+responsible for the health of the component and the component's community.
+The remainder of this specification will be focused on the details of
+the role.
+
+The term **core contributor** will mean a named person participating
+in the `Core Contributor Program`_ as a coding core contributor and
+meeting the requirements of that program.
+
+The term **maintainer scrum-of-scrums** will mean the group of all
+maintainers.  They are responsible for flagging issues to project
+leadership and proposing process improvements for component
+maintenance.
+
+The term **Technical Oversight Committee** (TOC) will mean the group
+responsible for the overall oversight and stewardship of the Open edX
+project. The TOC's function with respect to the project is to be the
+arbiter of last resort for governance matters. The TOC itself was established
+and is governed by a formal
+charter.  You can read the full document `here
+<https://openedx.org/wp-content/uploads/2021/12/TCRIL-Technical-Oversight-Committee-Charter-rev.11-16.21.pdf>`_.
+
+Historical Context
+------------------
+Historically, the Open edX project has not had a well-defined role
+of maintainer.  At times, community members stepped into this role
+informally.  The role was also part of a broader "owner" role at edX.
+Owners were also responsible for the health of the component they
+owned both in the repository and in production; owners answered the
+pager when something went wrong.
+
+What Maintainers Do
+-------------------
+The role of the repository or project maintainer is both a community
+focused and a technical role [#role]_.  Maintainers will be the public face
+of each component and should be welcoming and available.  While
+maintainers should strive to always be friendly and kind in their
+communication, firmness and directness are equally important.
+Maintainers must enforce the project's standards.  In cases where a
+proposed change isn't aligned with the our standards or direction, it
+is best to make that clear as early as possible.
+
+Non-technical responsibilities of maintainers include:
+
+* Effective communication with community members, other maintainers,
+  and project leadership
+* Selection and approval of Core Contributors with commit rights to
```

Comment:
[I've expressed these thoughts to others and was encouraged to post here!]

I'm a recent addition to edX/2U, and upon joining last year I was pleasantly surprised (and, honestly, terrified in a positive way) that I was able to merge commits to `master` in any repo.

I discovered that this was very empowering in several ways:
* I could offer manpower to perform deployments for other teams 
* I could revert other team's mistakes easily when on call
* I could unblock others by approving PRs for repos my team doesn't explicitly own but in which I happen to have an expertise in
* I could substitute for others when they were were unavailable
* (and vice-versa: people very often offer/revert/unblock/sub for me, and I sense this is a culture in the Open edX community in general!)

These capabilities are also vital given:
* How short-staffed edX/2U (and I'm sure everyone else) is. For example, two teams I've happened to have occasion to do the above for had and still have zero developers on team.
* How interconnected the Open edX codebase is. I have found bugs in my team's repo are caused by commits made 1-3 non-owned repos away. 

I am confused about the ramifications of this OEP to our ability to offer/revert/unblock/sub for others. This is because:

* I have heard mixed interpretations from others as to whether and to what extent this OEP will impact: (a) commit rights, (b) merge rights, and (c) PR approval requirements.
  * For example, in the most extreme case: Will the ability to create & commit to a new branch in a repo be on a repo-by-repo basis, as determined by the maintainer?
* I do not understand how this OEP and the Core Contributor OEP accommodate the fact that many maintainers are organizations or teams within organizations rather than individuals.
  * For example: Would a new employee of an organization be eligible to be a maintainer or even a Core Contributor? If not, would they be unable to fully ship code until they are eligible to become a Core Contributor?
* It seems like, at least initially, some repositories may need a very large number of maintainers. `edx-platform`, e.g.

Would someone be willing to help me gain clarity on this?

-------

A separate thought experiment: Why not add all Core Contributors to `pull-push-all` instead of restricting commit/approve/merge/deploy repo-by-repo?

Put another way:
* I don't understand why maintainers need a hard ACL control over which Core Contributors have commit privileges to their repo.
* I'd like to suggest allowing all Core Contributors to commit to all repos increases rather than decreases developer agency, for edX/2U but especially for the broader community.
   * (I'd appreciate input here especially if this is not right, as I'm not a member of the broader community.)

-------

I am sure I may be making a number of incorrect assumptions or am lacking some understanding due to error/skimming/newness/missing info/etc. I'm looking forward to hearing what they may be!

Reply:
@pshiu 's - I really appreciate your take here. You've made me step back and reflect on what the goals of our commit access control system should be. I can personally only think of four requirements:
1. Ensure that each contribution is reviewed by at least one person with the relevant context and then merged by a core contributor (who may or may not be the author).
2. Cause little-to-no undue friction for development and on-call responsibilities.
3. Provide a consistent & level playing field for the whole community.
4. Be easy to understand and administrate.

(Note: I didn't include anything about "quality" because it's so subjective and IMO will stem naturally from goals 1-3 in conjuction with the existing criteria for becoming a core contributor in the first place)

I do think the system we have been using for non-2U community members via the CC program could meet goals 1-4 if rolled out to 2U. That being said, the system you describe also meets those goals, are could be even stronger on goals 2 and 4!

As for goal 1 (ensuring review), perhaps that is where repo maintainers could make the judgement call on what "relevant context" is, using a tool like codeowners, main branch protection, or simple documentation to enforce review requirements.

I am curious what others think!

Reply:
Apologies in advance for this long response.  I would like to resolve this comment this week and be in a position to merge a **provisional** version of this OEP on Friday :slightly_smiling_face: 

Phil thanks for your response.  It’s clear that a lot of time and thought went into it.

I’d like to respond to a few of your points.  

Firstly, I don’t think that being able to “fully ship code” and “ship code instantly with only myself to decide” are the same thing.  Developer velocity is certainly a priority, but it’s not the only one.

Giving all core contributors universal, unfettered access is an interesting proposal.  However, I don’t think it’s the right choice.  It is standard practice in open source projects - and many closed source ones as well -  that rights to commit to or maintain the software are granted based on evidence.  Over time contributors demonstrate that they have expertise, are reliable, and have good judgment.

One of the critical jobs of the proposed maintainer role is to ensure that the software is of high quality.  I believe that we should be, for example, minimizing rollbacks rather than optimizing for them.  I do think we’ll need to have an affordance, say 2U admins with special privileges, to ensure that 2U operations remain stable and fast.  But 2U are not the only ones running the software in production.  So we’ll have to consider whether other organizations should also have such admins.  

Ultimately, I would like to see 2U have the same privileges as the rest of the community.  I think that it makes sense for this to happen by decreasing 2U’s default privileges, not by elevating everyone else’s to a point that I don’t think is safe or supported by best practices. We are in parallel working on a pilot of decreasing permissions (“scoped access permissions pilot”), separate from this OEP. This OEP does not impact the areas you mention: (a) commit rights, (b) merge rights, and (c) PR approval requirements. This OEP is about how to make sure all repos are maintained/have high quality standards.

2U – as other community members always have needed to do – will have to have some operational procedures in place that ensure safety, stability, and velocity.  I know that Ned among others has been thinking about this already. To your questions, “Would a new employee of an organization be eligible to be a maintainer or even a Core Contributor? If not, would they be unable to fully ship code until they are eligible to become a Core Contributor?” - this will be tackled with the scoped access permissions pilot. Let’s hold this conversation for then.

Regarding the complexity that is edx-platform, the OEP provides an affordance for teams as maintainers and also indicates that for repositories like edx-platform, parts of it can be maintained by distinct people or teams.

Reply:
Thanks for the response, @e0d! Like you said, I'm sure these concerns will be tackled in the scoped access permissions pilot, so looking forward to following the discussions there.